tmail-backend/jmap/extensions/src/main/java/com/linagora/tmail/james/jmap/firebase/FirebasePushClient.java

```diff
@@ -0,0 +1,69 @@
+package com.linagora.tmail.james.jmap.firebase;
+
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.firebase.FirebaseApp;
+import com.google.firebase.FirebaseOptions;
+import com.google.firebase.messaging.FirebaseMessaging;
+import com.google.firebase.messaging.Message;
+import com.linagora.tmail.james.jmap.model.FirebaseToken;
+import com.linagora.tmail.james.jmap.model.MissingOrInvalidFirebaseCredentialException;
+
+import reactor.core.publisher.Mono;
+
+public class FirebasePushClient {
+    private static final Logger LOGGER = LoggerFactory.getLogger(FirebasePushClient.class);
+
+    private final FirebaseMessaging firebaseMessaging;
+
+    @Inject
+    public FirebasePushClient(FirebaseConfiguration configuration) {
+        try {
+            FirebaseOptions options = FirebaseOptions.builder()
+                .setCredentials(GoogleCredentials.fromStream(new FileInputStream(configuration.firebasePrivateKeyUrl())))
+                .build();
+
+            FirebaseApp.initializeApp(options);
+        } catch (FileNotFoundException e) {
+            LOGGER.error("Declared private key file for Firebase is not found");
+            throw new MissingOrInvalidFirebaseCredentialException(e.getMessage());
+        } catch (IOException e) {
+            LOGGER.error("Declared private key file for Firebase is invalid");
+            throw new MissingOrInvalidFirebaseCredentialException(e.getMessage());
+        }
+
+        this.firebaseMessaging = FirebaseMessaging.getInstance();
+    }
+
+    public Mono<Void> push(FirebasePushRequest pushRequest) {
+        return Mono.fromCallable(() -> firebaseMessaging.send(createFcmMessage(pushRequest)))
+            .then();
+    }
+
+    public Mono<Void> validateToken(FirebaseToken token) {
+        return Mono.fromCallable(() -> firebaseMessaging.send(createDummyMessage(token), true))
+            .then();
+    }
+
+    private Message createFcmMessage(FirebasePushRequest pushRequest) {
+        return Message.builder()
+            .putAllData(pushRequest.stateChangesMap())
+            .setToken(pushRequest.token().value())
+            .build();
+    }
+
+    private Message createDummyMessage(FirebaseToken token) {
```

Comment:
Is this really a dummy message? Maybe we can find a better name for a request that is validating the token?